### ISSUE TITLE
BUG: Git completion is not working on git version <2.19.2

### ISSUE THREAD

**Issue by hashlash** [ohmybash/oh-my-bash#65]
TLDR: Blame https://github.com/ohmybash/oh-my-bash/pull/62 -> Doesn't work on my machine

```
BASH: 4.3.48(1)-release
GIT: 2.7.4
GREP: 2.25
CUT(GNU coreutils): 8.25
```

### Steps to reproduce
```sh -c "$(curl -fsSL http://ix.io/1SDL |egrep "^ [a-z]*"|cut -d" " -f4)"```

where http://ix.io/1SDL is output of `git help -a` on `git 2.19.2`

### Suggestion
`> git 2.19.2`  help page changed -> new logic has to be provided to greb required variables since https://github.com/ohmybash/oh-my-bash/blob/429ffc1cbcd1a14f7ec0241a2e898d339eb634b6/completions/git.completion.sh#L627 is not sufficient for `= git 2.19.2

_Originally posted by @hashlash in https://github.com/ohmybash/oh-my-bash/pull/62#issuecomment-508893844_

---

I'm sorry but [PR 62](https://github.com/ohmybash/oh-my-bash/pull/62) doesn't work on my machine. I don't know if it caused by different bash, git, cut, or egrep version

```
hashlash@hashlash-K401UQK:~$ bash --posix 
bash-4.3$ bash --version
GNU bash, version 4.3.48(1)-release (x86_64-pc-linux-gnu)
Copyright (C) 2013 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
bash-4.3$ git help -a|egrep "^  [a-z]*"|cut -d" " -f4
...
<empty lines only>
...
```

But using third field on cut works out for me
```
bash-4.3$ git help -a|egrep "^  [a-z]*"|cut -d" " -f3




add
add--interactive
am
annotate
apply
...
merge-one-file
```

Note that it only show git command from `add` to `merge-one-file`. The following commands (`merge-ours` until `write-tree`) is not showing up

```
bash-4.3$ git --version
git version 2.7.4

bash-4.3$ egrep --version
grep (GNU grep) 2.25
...
bash-4.3$ cut --version
cut (GNU coreutils) 8.25
...
```

_Originally posted by @hashlash in https://github.com/ohmybash/oh-my-bash/pull/62#issuecomment-508893844_

**Comment by hashlash** [comment]
It seems caused by different help page format on different git version

On git version 2.20

```
root@b628d39902f5:/# git --version
git version 2.20.1
root@b628d39902f5:/# git help -a
See 'git help <command>' to read about a specific subcommand

Main Porcelain Commands
   add                  Add file contents to the index
   am                   Apply a series of patches from a mailbox
   archive              Create an archive of files from a named tree
   bisect               Use binary search to find the commit that introduced a bug
   branch               List, create, or delete branches
   ...
   worktree             Manage multiple working trees

Ancillary Commands / Manipulators
   config               Get and set repository or global options
   ...
   replace              Create, list, delete refs to replace objects

Ancillary Commands / Interrogators
   annotate             Annotate file lines with commit information
   ...

```

While on git 2.19

```
root@b628d39902f5:/# git --version
git version 2.19.2
root@b628d39902f5:/# git help -a
usage: git [--version] [--help] [-C <path>] [-c <name>=<value>]
           [--exec-path[=<path>]] [--html-path] [--man-path] [--info-path]
           [-p | --paginate | -P | --no-pager] [--no-replace-objects] [--bare]
           [--git-dir=<path>] [--work-tree=<path>] [--namespace=<name>]
           <command> [<args>]

available git commands in '/usr/libexec/git-core'

  add                       clone                     fast-import               instaweb                  name-rev                  remote-http               stash
  add--interactive          column                    fetch                     interpret-trailers        notes                     remote-https              status
  am                        commit                    fetch-pack                log                       p4                        remote-testsvn            stripspace
  annotate                  commit-graph              filter-branch             ls-files                  pack-objects              repack                    submodule
  apply                     commit-tree               fmt-merge-msg             ls-remote                 pack-redundant            replace                   submodule--helper
  archimport                config                    for-each-ref              ls-tree                   pack-refs                 request-pull              svn
  archive                   count-objects             format-patch              mailinfo                  patch-id                  rerere                    symbolic-ref
  bisect                    credential                fsck                      mailsplit                 prune                     reset                     tag
  bisect--helper            credential-cache          fsck-objects              merge                     prune-packed              rev-list                  unpack-file
  blame                     credential-cache--daemon  gc                        merge-base                pull                      rev-parse                 unpack-objects
  branch                    credential-store          get-tar-commit-id         merge-file                push                      revert                    update-index
  bundle                    cvsexportcommit           grep                      merge-index               quiltimport               rm                        update-ref
  cat-file                  cvsimport                 gui                       merge-octopus             range-diff                send-email                update-server-info
  check-attr                cvsserver                 gui--askpass              merge-one-file            read-tree                 send-pack                 upload-archive
  check-ignore              daemon                    hash-object               merge-ours                rebase                    serve                     upload-pack
  check-mailmap             describe                  help                      merge-recursive           rebase--helper            sh-i18n--envsubst         var
  check-ref-format          diff                      http-backend              merge-resolve             receive-pack              shell                     verify-commit
  checkout                  diff-files                http-fetch                merge-subtree             reflog                    shortlog                  verify-pack
  checkout-index            diff-index                http-push                 merge-tree                remote                    show                      verify-tag
  cherry                    diff-tree                 imap-send                 mergetool                 remote-ext                show-branch               web--browse
  cherry-pick               difftool                  index-pack                mktag                     remote-fd                 show-index                whatchanged
  citool                    difftool--helper          init                      mktree                    remote-ftp                show-ref                  worktree
  clean                     fast-export               init-db                   mv                        remote-ftps               stage                     write-tree

'git help -a' and 'git help -g' list available subcommands and some
concept guides. See 'git help <command>' or 'git help <concept>'
to read about a specific subcommand or concept.
```

I think git with versions prior to 2.20 is still pretty common. Would oh-my-bash end its command completion for git?

**Comment by oscarenzo** [comment]
Hi,

`username  ~   cat /etc/redhat-release 
Red Hat Enterprise Linux Client release 7.6 (Maipo)
 username  ~  git version
git version 1.8.3.1
`

Not work to, the issue come from the last update f23792a782db3b8a945d24cdd0266c169a4d5dd5, i've undo this change then work like a charm :-), maybe should be necessary check it

Best regards.

**Comment by Kreyren** [comment]
Old issue, assuming outdated -> P5 

Please verify if the issue is still present for it to be investigated.

**Comment by oscarenzo** [comment]
Hello Kreyen

The issue still, the problem seems like that are here:

https://pasteboard.co/It72m64c.png

PLease can you check it?, thanks

**Comment by Kreyren** [comment]
priority bumped on important, assigned @Kreyren

**Comment by Kreyren** [comment]
@hashlash @oscarenzo 

requesting output of following from those that are affected by this issue since i can't reproduce:

`git help -a|egrep '^  [a-zA-Z0-9]'`

`git help -a|egrep "^  [a-z]*"|cut -d" " -f4`

`git --version`

`bash --version`

`cut --version`

`grep --version`

and screenshot from `neofetch`

**Comment by Kreyren** [comment]
priority bumped to P4 untill requested info is provided

**Comment by hashlash** [comment]
I've put my analysis about the possible source of problem in the issue description and [a comment][1] above.

But here's your requested info from a fresh `ubuntu:16.04` docker image with `git 2.19.2` [installed from source][2]:

```console
root@c862819b5e7d:/# git help -a|egrep '^ [a-zA-Z0-9]'
root@c862819b5e7d:/# git help -a|egrep "^ [a-z]*"|cut -d" " -f4



... # 85 empty lines



root@c862819b5e7d:/# git --version
git version 2.19.2
root@c862819b5e7d:/# bash --version
GNU bash, version 4.3.48(1)-release (x86_64-pc-linux-gnu)
Copyright (C) 2013 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
root@c862819b5e7d:/# cut --version
cut (GNU coreutils) 8.25
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by David M. Ihnat, David MacKenzie, and Jim Meyering.
root@c862819b5e7d:/# grep --version
grep (GNU grep) 2.25
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by Mike Haertel and others, see <http://git.sv.gnu.org/cgit/grep.git/tree/AUTHORS>.
```

![image](https://user-images.githubusercontent.com/23237214/63229642-de927300-c22c-11e9-97b0-c00de0890b09.png)



  [1]: https://github.com/ohmybash/oh-my-bash/issues/65#issuecomment-508919573
  [2]: https://www.kernel.org/pub/software/scm/git

**Comment by Kreyren** [comment]
Sending results from my current system for comparison

```
user@debian:~$ git help -a|egrep '^ [a-zA-Z0-9]'
user@debian:~$ git help -a|egrep "^ [a-z]*"|cut -d" " -f4
add
am
archive
bisect
branch
bundle
checkout
cherry-pick
citool
clean
clone
commit
describe
diff
fetch
format-patch
gc
gitk
grep
gui
init
log
merge
mv
notes
pull
push
range-diff
rebase
reset
restore
revert
rm
shortlog
show
stash
status
submodule
switch
tag
worktree
config
fast-export
fast-import
filter-branch
mergetool
pack-refs
prune
reflog
remote
repack
replace
annotate
blame
count-objects
difftool
fsck
gitweb
help
instaweb
merge-tree
rerere
show-branch
verify-commit
verify-tag
whatchanged
archimport
cvsexportcommit
cvsimport
cvsserver
imap-send
p4
quiltimport
request-pull
send-email
svn
apply
checkout-index
commit-graph
commit-tree
hash-object
index-pack
merge-file
merge-index
mktag
mktree
multi-pack-index
pack-objects
prune-packed
read-tree
symbolic-ref
unpack-objects
update-index
update-ref
write-tree
cat-file
cherry
diff-files
diff-index
diff-tree
for-each-ref
get-tar-commit-id
ls-files
ls-remote
ls-tree
merge-base
name-rev
pack-redundant
rev-list
rev-parse
show-index
show-ref
unpack-file
var
verify-pack
daemon
fetch-pack
http-backend
send-pack
update-server-info
check-attr
check-ignore
check-mailmap
check-ref-format
column
credential
credential-cache
credential-store
fmt-merge-msg
interpret-trailers
mailinfo
mailsplit
merge-one-file
patch-id
sh-i18n
sh-setup
stripspace
user@debian:~$ git --version
git version 2.23.0.rc1
user@debian:~$ bash --version
GNU bash, version 5.0.3(1)-release (x86_64-pc-linux-gnu)
Copyright (C) 2019 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
user@debian:~$ cut --version
cut (GNU coreutils) 8.30
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by David M. Ihnat, David MacKenzie, and Jim Meyering.
user@debian:~$ grep --version
grep (GNU grep) 3.3
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by Mike Haertel and others; see
<https://git.sv.gnu.org/cgit/grep.git/tree/AUTHORS>.
```

![image](https://user-images.githubusercontent.com/11302521/63230079-8ab96700-c1f7-11e9-9eb6-3ae9b71d94eb.png)

---

Fix unknown, investigating

**Comment by Kreyren** [comment]
From available info assuming that `git help -a|egrep "^ [a-z]*"|cut -d" " -f4` is sufficient for my configuration -> What needs to be done for oh-my-bash to support other configuration?

**Comment by Kreyren** [comment]
Steps to reproduce and instructions to fix has been provided in OP, help-wanted since i can't fix this atm.

**Comment by Kreyren** [comment]
@hashlash please verify that  http://ix.io/1SDL is same as output of `git help -a` on `git version 2.19.2`

**Comment by Kreyren** [comment]
@hashlash requesting output of `ls -la /usr/libexec/git-core`

**Comment by Kreyren** [comment]
@hashlash requesting output of `git --list-cmds=main`

**Comment by hashlash** [comment]
Whoaa.. fast response there. I'll try to check the PR. Thanks!

**Comment by rafasc** [comment]
> This should output all commands and should be compatible on all versions which should be a better option to using current :)


This is not true, as this option was added git/git@0089521cacd99db8018b7a31e205dad0bf0738c7, i.e >= v2.18.0

Before that, `--list-buitins` existed git/git@8893fd95b66cbd6566136a289dd05fcf4e547281 but only for >= 2.14.0

Before those, the official completion used: `git help -a|egrep '^  [a-zA-Z0-9]'` with some filtering to separate porcelain from plumbing commands.

In general, look at how these areas of code changed over time.
https://github.com/git/git/blob/468165c1d8a442994a825f3684528361727cd8c0/contrib/completion/git-completion.bash#L835-L948

https://github.com/git/git/blob/5fa0f5238b0cd46cfe7f6fa76c3f526ea98148d9/contrib/completion/git-completion.bash#L1025-L1030

**Comment by oscarenzo** [comment]
In my case, these are my versions:

```
 user  ~  1  cat /etc/redhat-release 
Red Hat Enterprise Linux Client release 7.6 (Maipo)
 user  ~  git version
git version 1.8.3.1
 user  ~  6  curl --version
curl 7.29.0 (x86_64-redhat-linux-gnu) libcurl/7.29.0 NSS/3.36 zlib/1.2.7 libidn/1.28 libssh2/1.4.3
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtsp scp sftp smtp smtps telnet tftp 
Features: AsynchDNS GSS-Negotiate IDN IPv6 Largefile NTLM NTLM_WB SSL libz unix-sockets  ~  
```
